packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -568,18 +593,89 @@ function updateHostText(current, workInProgress) {
   return null;
 }
 
+function resolveDefaultProps(Component, baseProps) {
+  if (Component && Component.defaultProps) {
+    // Resolve default props. Taken from ReactElement
+    const props = Object.assign({}, baseProps);
+    const defaultProps = Component.defaultProps;
+    for (let propName in defaultProps) {
+      if (props[propName] === undefined) {
+        props[propName] = defaultProps[propName];
+      }
+    }
+    return props;
+  }
+  return baseProps;
+}
+
 function mountIndeterminateComponent(
   current,
   workInProgress,
+  Component,
   renderExpirationTime,
 ) {
   invariant(
     current === null,
     'An indeterminate component should never have mounted. This error is ' +
       'likely caused by a bug in React. Please file an issue.',
   );
-  const fn = workInProgress.type;
+
   const props = workInProgress.pendingProps;
+  if (
+    typeof Component === 'object' &&
+    Component !== null &&
+    typeof Component.then === 'function'
+  ) {
+    Component = readLazyComponentType(Component);
+    reassignLazyComponentTag(workInProgress, Component);
+    const resolvedProps = resolveDefaultProps(Component, props);
+    let child;
+    switch (workInProgress.tag) {
+      case FunctionalComponentLazy: {
+        child = updateFunctionalComponent(
+          current,
+          workInProgress,
+          Component,
+          resolvedProps,
+          renderExpirationTime,
+        );
+        break;
+      }
+      case ClassComponentLazy: {
+        child = updateClassComponent(
+          current,
+          workInProgress,
+          Component,
+          resolvedProps,
+          renderExpirationTime,
+        );
+        break;
+      }
+      case ForwardRefLazy: {
+        child = updateForwardRef(
+          current,
+          workInProgress,
+          Component,
+          resolvedProps,
+          renderExpirationTime,
+        );
+        break;
+      }
+      default: {
+        // This message intentionally doesn't metion ForwardRef because the
+        // fact that it's a separate type of work is an implementation detail.
+        invariant(
+          false,
+          'Element type is invalid. Received a promise that resolves to: %s. ' +
+            'Promise elements must resolve to a class or function.',
+          Component,
+        );
+      }
+    }
+    workInProgress.pendingProps = props;
```

Comment:
Why is this needed?

Reply:
The pending props needs to match the memoized props so that the next time we visit this fiber, when we compare `memoizedProps` to `pendingProps`, they are the same object.

Although this looks like a mistake. I think it should be `workInProgress.pendingProps = resolvedProps`. Clearly I forgot to write a test... will do.

Reply:
That doesn't make sense because memoizedProps on other paths stores the unresolved props. I think it needs to store the unresolved props because if you store the resolved ones, you'll never get reference equality.

Reply:
Nope, I did write a test. But if I delete it, it passes. I think it was failing in an earlier version.

It's not needed because we compare to current's `pendingProps` instead of `memoizedProps`. Which kinda suggests we should delete one of those fields :D